src/mcp/shared/session.py

```diff
@@ -204,13 +225,34 @@ async def _send_response(
             )
             await self._write_stream.send(JSONRPCMessage(jsonrpc_response))
 
+    async def _cleanup_loop(self) -> None:
+        """Periodically clean up completed and cancelled requests."""
+        while True:
+            with anyio.move_on_after(self._cleanup_interval):
+                # Clean up completed requests
+                self._in_flight = {
+                    req_id: responder
+                    for req_id, responder in self._in_flight.items()
+                    if responder.in_flight
+                }
+                await anyio.sleep(self._cleanup_interval)
+
     async def _receive_loop(self) -> None:
+        """Handle incoming messages and maintain request state."""
+
         async with (
             self._read_stream,
             self._write_stream,
             self._incoming_message_stream_writer,
         ):
             async for message in self._read_stream:
+                # Clean up completed requests
+                self._in_flight = {
+                    req_id: responder
+                    for req_id, responder in self._in_flight.items()
+                    if responder.in_flight
+                }
```

Comment:
OTOH, this seems excessive to do on every new incoming message. Why can't we tie it to completion and cancellation?